stellargraph/connector/neo4j/graph.py

```diff
@@ -58,25 +66,94 @@ def nodes(self):
+
+    def _node_features_from_db(self, nodes):
+        if nodes is None:
+            where_clause = ""
+            valid = slice(None)
+        else:
+            where_clause = " WHERE node.ID = node_id"
+            if isinstance(nodes, np.ndarray):
+                valid = nodes != None
+                # we need to create a list of integers to run the neo4j query with
+                nodes = [int(node) for node in nodes]
+            elif isinstance(nodes, list):
+                valid = np.array(nodes) != None
+            else:
+                # single node
+                valid = np.array([nodes is not None])
+                nodes = [nodes]
+
+        # None's should be filled with zeros in the feature matrix
+        features = np.zeros((len(nodes), self.node_feature_sizes()[self._node_type]))
+
+        # fill valid locs with features
+        feature_query = f"""
+            UNWIND $node_id_list AS node_id
+            MATCH(node) {where_clause}
+            RETURN node.features as features
+            """
+        result = self.graph_db.run(feature_query, parameters={"node_id_list": nodes})
+        rows = result.data()
+
+        # this method currently doesn't handle any other invalid IDs. If there are other
+        # invalid ids, we should raise an error
+        if len(rows) != sum(valid):
+            raise ValueError(
+                "nodes: Found values that did not return any results from the database."
+            )
+
+        features[valid, :] = np.array([row["features"] for row in rows])
+
+        return features
+
+    def node_features(self, nodes):
         """
         Get the numeric feature vectors for the specified nodes or node type.
 
         Args:
-            nodes (list): list of node IDs.
+            nodes (list or hashable, optional): Node ID or list of node IDs.
         Returns:
             Numpy array containing the node features for the requested nodes.
         """
-        feature_query = f"""
-            UNWIND $node_id_list AS node_id
-            MATCH(node) WHERE node.ID = node_id
-            RETURN node.features as features
-            """
-        result = self.graph_db.run(
-            feature_query, parameters={"node_id_list": node_ids},
-        )
-        features = np.array([row["features"] for row in result.data()])
-        return features
+        if self._nodes is not None:
+            return _features(
+                self._nodes,
+                self.unique_node_type,
+                "nodes",
+                nodes,
+                type=None,
+                use_ilocs=False,
+            )
+        else:
+            return self._node_features_from_db(nodes)
+
+    def unique_node_type(self):
+        return self._node_type
+
+    def node_feature_sizes(self):
+        if self._node_feature_size is None:
+            # if feature size is unknown, take a random node's features
```

Comment:
This is a little concerning to me, because it's pushing us towards having to handle cache invalidation and similar things like that. Maybe we could at least write down the assumptions a bit better, e.g. off the top of my head:

- all nodes have features as a single list (under the `features` property, but I see #1592)
- all nodes' features have the same size

Reply:
I've written down the assumptions in the docstring - what do you think?

Reply:
One extra addition, but otherwise great!